onnxruntime/core/providers/qnn/builder/opbuilder/simple_op_builder.cc

```diff
@@ -163,7 +209,7 @@ Status SimpleOpBuilder::ProcessAttributesAndOutputs(QnnModelWrapper& qnn_model_w
   if (do_op_validation) {
     ORT_RETURN_IF_ERROR(ExplictOpCheck(qnn_model_wrapper, node_unit));
     // Skip the op validation for DepthToSpace & SpaceToDepth if it's not NHWC data layout
-    if (node_unit.Domain() != kMSInternalNHWCDomain && (op_type == "DepthToSpace" || op_type == "SpaceToDepth")) {
+    if (node_unit.Domain() != kMSInternalNHWCDomain && (op_type == "DepthToSpace" || op_type == "SpaceToDepth" || op_type == "GridSample")) {
```

Comment:
I believe that skipping the validation for these operators (when not in NHWC) can lead to an irrecoverable error in the graph partitioner.

For example, if a GridSample operator uses an unsupported `mode` attribute (e.g., cubic), the first call to GetCapability will incorrectly accept this op because we're skipping attribute validation. On the second call to GetCapability (after nhwc layout transformation), the validation will reject the op due to the unsupported mode. This will cause the graph partitioner to enter an invalid state with an error such as:

> Node 'GridSample' OpType:GridSample with domain:com.ms.internal.nhwc was inserted using the NHWC format as requested by QNNExecutionProvider, but was not selected by that EP. This means the graph is now invalid as there will not be an EP able to run the node. This could be a bug in layout transformer, or in the GetCapability implementation of the EP.
failed to initialize onnx model

We've seen this error before with Resize ops. I believe that the SimpleOpBuilder should override the `IsOpSupported` virtual method allow at least basic attribute validation.

Reply:
There are logic to check the mode and padding_mode in ExplictOpCheck by re-using the ArrayHasString method you provided. 

---
In reply to: [1309080120](https://github.com/microsoft/onnxruntime/pull/17317#discussion_r1309080120) [](ancestors = 1309080120)

Reply:
Ah yes, you're right. Thanks.

Reply:
Where did you get that error message for GridSample?
Yes, if we failed to filter out nodes before the layout transformer applied. it will cause issue we filter it out after layout transformer. There's no good solution for that right now.

---
In reply to: [1309080120](https://github.com/microsoft/onnxruntime/pull/17317#discussion_r1309080120) [](ancestors = 1309080120)